Python/Product/Analysis/LanguageServer/Server.cs

```diff
@@ -622,49 +611,45 @@ private sealed class ReloadModulesQueueItem : IAnalyzable {
-            var severity = _settings?.diagnostics?.unresolvedImports ?? DiagnosticSeverity.Warning;
-            if (parse != null && severity.Show()) {
-                var walker = new ImportStatementWalker(parse.Tree, pythonProjectEntry, _analyzer, severity);
-                parse.Tree.Walk(walker);
-                diags = diags.Concat(walker.Diagnostics).ToArray();
-            }
-
-            if (!diags.Any()) {
-                return;
-            }
-
-            if (pythonProjectEntry is IDocument doc) {
-                var reported = _openFiles.GetDocument(doc.DocumentUri).LastReportedDiagnostics;
-                lock (reported) {
-                    if (reported.TryGetValue(0, out var lastVersion)) {
-                        diags = diags.Concat(lastVersion.Diagnostics).ToArray();
-                    }
-                }
-            }
-
-            PublishDiagnostics(new PublishDiagnosticsEventArgs {
-                diagnostics = diags,
-                uri = pythonProjectEntry.DocumentUri,
-                _version = version
-            });
+            _editorFiles.GetDocument(pythonProjectEntry.DocumentUri).UpdateAnalysisDiagnostics(pythonProjectEntry, version);
         }
 
         private PythonAst GetParseTree(IPythonProjectEntry entry, Uri documentUri, CancellationToken token, out BufferVersion bufferVersion) {
             PythonAst tree = null;
             bufferVersion = null;
-            var parse = entry.WaitForCurrentParse(_clientCaps?.python?.completionsTimeout ?? Timeout.Infinite, token);
+            var parse = entry.WaitForCurrentParse(Timeout.Infinite, token);
             if (parse != null) {
                 tree = parse.Tree;
                 if (parse.Cookie is VersionCookie vc) {
-                    if (vc.Versions.TryGetValue(_projectFiles.GetPart(documentUri), out bufferVersion)) {
+                    if (vc.Versions.TryGetValue(ProjectFiles.GetPart(documentUri), out bufferVersion)) {
                         tree = bufferVersion.Ast ?? tree;
                     }
                 }
             }
             return tree;
         }
 
+        private bool HandleConfigurationChanges(LanguageServerSettings newSettings) {
+            var oldSettings = Settings;
+            Settings = newSettings;
+
+            if (oldSettings == null) {
+                return true;
+            }
+
+            if (newSettings.analysis.openFilesOnly != oldSettings.analysis.openFilesOnly) {
+                _editorFiles.UpdateDiagnostics();
+                return false;
+            }
+
+            if (!newSettings.analysis.errors.Intersect(oldSettings.analysis.errors).Any() ||
+                !newSettings.analysis.warnings.Intersect(oldSettings.analysis.warnings).Any() ||
+                !newSettings.analysis.information.Intersect(oldSettings.analysis.information).Any() ||
+                !newSettings.analysis.disabled.Intersect(oldSettings.analysis.disabled).Any()) {
```

Comment:
This isn't the right check either - will skip updating when changing from `['a', 'b']` to `['b', 'c']`.

I wish Linq had a `.SetEqual` extension method like `.SequenceEqual` (feel free to add it to Analysis.Infrastructure if you want), but it really does have to look like:

```
if (!new HashSet<string>(newSetting.analysis.errors).SetEquals(oldSettings.analysis.errors)) { ... }
```

Reply:
OK